extensions/amp-web-push/0.1/web-push-service.js

```diff
@@ -0,0 +1,697 @@
+      */
+      const isSecureUrl = (parsedUrl.protocol === 'https:');
+
+      return isSecureUrl && isNotRootUrl;
+    } catch (e) {
+      return false;
+    }
+  }
+
+  /*
+   * Wait for bind scan to finish for testing.
+   *
+   * @return {?Promise}
+   * @visibleForTesting
+   */
+  get initializePromiseForTesting() {
+    return this.initializePromise_;
+  }
+
+  /*
+    Waits until the helper iframe has loaded, and then sends the message to the
+    helper iframe and awaits a reply. Errors that are returned are thrown,
+    otherwise the message is returned as a Promise.
+
+    This is used by all of our AMP page <-> helper iframe communications.
+   */
+  queryHelperFrame_(messageTopic, message) {
+    return this.iframe_.whenReady().then(() => {
+      return this.frameMessenger_.send(messageTopic, message);
+    }).then(result => {
+      const replyPayload = result[0];
+      if (replyPayload.success) {
+        return replyPayload.result;
+      } else {
+        throw new Error(`AMP page helper iframe query topic ${messageTopic} ` +
+          `and message ${message} failed with: ${replyPayload.error}`);
+      }
+    });
+  }
+
+  /*
+    Passes messages to the service worker through the helper iframe. Messages
+    are forwarded directly as-is and service worker replies and received as-is
+    without filtering, so that changes in the AMP page and service worker don't
+    require code changes in the helper frame (which lives on the canonical
+    origin).
+   */
+  queryServiceWorker_(messageTopic, message) {
+    return this.queryHelperFrame_(
+        WindowMessenger.Topics.SERVICE_WORKER_QUERY,
+        {
+          topic: messageTopic,
+          payload: message,
+        }
+    );
+  }
+
+  queryNotificationPermission_() {
+    return this.queryHelperFrame_(
+        WindowMessenger.Topics.NOTIFICATION_PERMISSION_STATE,
+        null
+    );
+  }
+
+  queryServiceWorkerState_() {
+    return this.queryHelperFrame_(
+        WindowMessenger.Topics.SERVICE_WORKER_STATE,
+        null
+    );
+  }
+
+  registerServiceWorker() {
+    return this.queryHelperFrame_(
+        WindowMessenger.Topics.SERVICE_WORKER_REGISTRATION,
+        {
+          workerUrl: this.config_.serviceWorkerUrl,
+          registrationOptions: this.config_.serviceWorkerRegistrationOptions ||
+          {scope: '/'},
+        }
+    );
+  }
+
+  querySubscriptionStateRemotely_() {
+    return this.queryServiceWorker_(
+        'amp-web-push-subscription-state',
+        null
+    );
+  }
+
+  subscribeForPushRemotely_() {
+    return this.queryServiceWorker_(
+        'amp-web-push-subscribe',
+        null
+    );
+  }
+
+  unsubscribeFromPushRemotely_() {
+    return this.queryServiceWorker_(
+        'amp-web-push-unsubscribe',
+        null
+    );
+  }
+
+  isServiceWorkerActivated_() {
+    const self = this;
+    return this.queryServiceWorkerState_().then(
+      /** @param {{isControllingFrame:boolean, state:string, url:string}}
```

Comment:
Ideally this would be defined in the `queryServiceWorkerState_`'s jsdoc instead of here.